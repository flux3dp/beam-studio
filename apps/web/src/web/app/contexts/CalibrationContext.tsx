import React, { createContext, useState } from 'react';

import DeviceMaster from 'helpers/device-master';
import PreviewModeController from 'app/actions/beambox/preview-mode-controller';
import storage from 'implementations/storage';
import versionChecker from 'helpers/version-checker';
import { CameraConfig } from 'interfaces/Camera';
import {
  DEFAULT_CAMERA_OFFSET,
  STEP_ASK_READJUST,
  STEP_PUT_PAPER,
} from 'app/constants/camera-calibration-constants';
import { IDeviceInfo } from 'interfaces/IDevice';

interface CalibrationContextType {
  borderless: boolean;
  cameraPosition: { x: number; y: number };
  setCameraPosition: (val: { x: number; y: number }) => void;
  calibratedMachines: string[];
  setCalibratedMachines: (val: string[]) => void;
  currentOffset: CameraConfig;
  setCurrentOffset: (val: CameraConfig) => void;
  currentStep: symbol;
  setCurrentStep: (val: symbol) => void;
  lastConfig: CameraConfig;
  setLastConfig: (val: CameraConfig) => void;
  gotoNextStep: (step: symbol) => void;
  onClose: (completed: boolean) => void;
  originFanSpeed: number;
  setOriginFanSpeed: (val: number) => void;
  imgBlobUrl: string;
  setImgBlobUrl: (v: string) => void;
  device: IDeviceInfo;
  unit: string;
}

export const CalibrationContext = createContext<CalibrationContextType>({
  borderless: false,
  cameraPosition: { x: 0, y: 0 },
  setCameraPosition: () => {},
  calibratedMachines: [],
  setCalibratedMachines: () => {},
  currentOffset: null,
  setCurrentOffset: () => {},
  currentStep: STEP_PUT_PAPER,
  setCurrentStep: () => {},
  lastConfig: null,
  setLastConfig: () => {},
  gotoNextStep: () => {},
  onClose: () => {},
  imgBlobUrl: '',
  setImgBlobUrl: () => {},
  originFanSpeed: 100,
  setOriginFanSpeed: () => {},
  device: null,
  unit: 'mm',
});

interface CalibrationProviderProps {
  children: React.ReactNode;
  borderless: boolean;
  device: IDeviceInfo;
  onClose: (completed: boolean) => void;
}

let calibratedMachines = [];

export function CalibrationProvider({
  children,
  borderless,
  device,
  onClose,
}: CalibrationProviderProps): JSX.Element {
  const didCalibrate = calibratedMachines.includes(device.uuid);
  const initStep: symbol = didCalibrate ? STEP_ASK_READJUST : STEP_PUT_PAPER;
  const [currentStep, setCurrentStep] = useState(initStep);
  const [currentOffset, setCurrentOffset] = useState(DEFAULT_CAMERA_OFFSET);
  const [lastConfig, setLastConfig] = useState(DEFAULT_CAMERA_OFFSET);
  const setCalibratedMachines = (newList: string[]) => {
    calibratedMachines = newList;
  };
  const [imgBlobUrl, setImgBlobUrl] = useState('');
  const [cameraPosition, setCameraPosition] = useState({ x: 0, y: 0 });
  const [originFanSpeed, setOriginFanSpeed] = useState(1000);
  const unit = (storage.get('default-units') as string) || 'mm';

  const wrappedOnClose = async (completed: boolean) => {
    onClose(completed);
    await PreviewModeController.end({ shouldWaitForEnd: true });
    const tempCmdAvailable = versionChecker(device?.version).meetRequirement('TEMP_I2C_CMD');
    if (originFanSpeed && !tempCmdAvailable) {
      await DeviceMaster.setFan(originFanSpeed);
    }
  };

  return (
    <CalibrationContext.Provider
      value={{
        borderless,
        device,
        currentStep,
        setCurrentStep,
        cameraPosition,
        setCameraPosition,
        currentOffset,
        setCurrentOffset,
        lastConfig,
        setLastConfig,
        calibratedMachines,
        setCalibratedMachines,
        imgBlobUrl,
        setImgBlobUrl,
        originFanSpeed,
        setOriginFanSpeed,
        gotoNextStep: setCurrentStep,
        unit,
        onClose: wrappedOnClose,
      }}
    >
      {children}
    </CalibrationContext.Provider>
  );
}
