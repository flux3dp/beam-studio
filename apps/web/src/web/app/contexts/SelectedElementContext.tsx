import React, { createContext, memo, useCallback, useEffect, useState, useRef } from 'react';

import eventEmitterFactory from 'helpers/eventEmitterFactory';

const canvasEventEmitter = eventEmitterFactory.createEventEmitter('canvas');

export interface SelectedElementContextType {
  selectedElement: Element | null;
}

export const SelectedElementContext = createContext<SelectedElementContextType>({
  selectedElement: null,
});

interface Props {
  children: React.ReactNode;
}

export const SelectedElementContextProvider = memo(({ children }: Props): JSX.Element => {
  const [selectedElement, setSelectedElement] = useState<Element | null>(null);
  const selectedElementRef = useRef<Element | null>(null);

  const handleSetSelectedElem = useCallback(
    (elem: Element): void => {
      if (elem === selectedElementRef.current) return;
      selectedElementRef.current = elem;
      setSelectedElement((cur) => {
        if (cur === elem) return cur;
        (document.activeElement as HTMLInputElement).blur();
        return elem;
      });
    },
    []
  );

  useEffect(() => {
    canvasEventEmitter.on('SET_SELECTED_ELEMENT', handleSetSelectedElem);
    return () => {
      canvasEventEmitter.off('SET_SELECTED_ELEMENT', handleSetSelectedElem);
    };
  }, [handleSetSelectedElem]);

  return (
    <SelectedElementContext.Provider value={{ selectedElement }}>
      {children}
    </SelectedElementContext.Provider>
  );
});

export default {
  SelectedElementContext,
  SelectedElementContextProvider,
};
