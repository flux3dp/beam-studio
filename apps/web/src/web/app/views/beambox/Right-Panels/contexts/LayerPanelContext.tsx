import React, { createContext, useEffect, useState, useCallback, useMemo } from 'react';

import doLayersContainsVector from 'helpers/layer/check-vector';
import eventEmitterFactory from 'helpers/eventEmitterFactory';
import useForceUpdate from 'helpers/use-force-update';
import useWorkarea from 'helpers/hooks/useWorkarea';
import { getLayerElementByName } from 'helpers/layer/layer-helper';
import { promarkModels } from 'app/actions/beambox/constant';

interface ILayerPanelContext {
  selectedLayers: string[];
  setSelectedLayers: (selectedLayers: string[]) => void;
  forceUpdate: () => void;
  forceUpdateSelectedLayers: () => void;
  hasVector: boolean;
  hasGradient: boolean;
}

export const LayerPanelContext = createContext<ILayerPanelContext>({
  selectedLayers: [],
  setSelectedLayers: () => {},
  forceUpdate: () => {},
  forceUpdateSelectedLayers: () => {},
  hasVector: false,
  hasGradient: false,
});
const layerPanelEventEmitter = eventEmitterFactory.createEventEmitter('layer-panel');

interface Props {
  children?: React.ReactNode;
}

export const LayerPanelContextProvider = ({ children }: Props): JSX.Element => {
  const [hasVector, setHasVector] = useState<boolean>(false);
  const [hasGradient, setHasGradient] = useState<boolean>(false);
  const [selectedLayers, setSelectedLayers] = useState<string[]>([]);
  const forceUpdate = useForceUpdate();
  const forceUpdateSelectedLayers = useCallback(
    () => setSelectedLayers([...selectedLayers]),
    [selectedLayers]
  );
  const workarea = useWorkarea();
  const isPromark = useMemo(() => promarkModels.has(workarea), [workarea]);

  const lazySetSelectedLayers = useCallback(
    (newLayers: string[]) => {
      if (
        newLayers.length === selectedLayers.length &&
        newLayers.every((name, i) => name === selectedLayers[i])
      ) {
        return;
      }
      setSelectedLayers(newLayers);
    },
    [selectedLayers, setSelectedLayers]
  );

  useEffect(() => {
    layerPanelEventEmitter.on('UPDATE_LAYER_PANEL', forceUpdate);
    return () => {
      layerPanelEventEmitter.removeListener('UPDATE_LAYER_PANEL', forceUpdate);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    layerPanelEventEmitter.on('SET_SELECTED_LAYERS', lazySetSelectedLayers);
    return () => {
      layerPanelEventEmitter.removeListener('SET_SELECTED_LAYERS', lazySetSelectedLayers);
    };
  }, [lazySetSelectedLayers]);

  useEffect(() => {
    const getSelectedLayers = (response: { selectedLayers: string[] }) => {
      response.selectedLayers = selectedLayers;
    };
    layerPanelEventEmitter.on('GET_SELECTED_LAYERS', getSelectedLayers);
    const checkVector = () => {
      const newVal = doLayersContainsVector(selectedLayers);
      setHasVector(newVal);
    };
    layerPanelEventEmitter.on('CHECK_VECTOR', checkVector);
    return () => {
      layerPanelEventEmitter.removeListener('GET_SELECTED_LAYERS', getSelectedLayers);
      layerPanelEventEmitter.removeListener('CHECK_VECTOR', checkVector);
    };
  }, [selectedLayers]);

  useEffect(() => {
    const checkGradient = () => {
      if (isPromark) {
        const newVal = selectedLayers.some((layerName: string) => {
          const layer = getLayerElementByName(layerName);
          return !!layer?.querySelector('image[data-shading="true"]');
        });
        setHasGradient(newVal);
      }
    };
    checkGradient();
    layerPanelEventEmitter.on('CHECK_GRADIENT', checkGradient);
    return () => {
      layerPanelEventEmitter.removeListener('CHECK_GRADIENT', checkGradient);
    };
  }, [selectedLayers, isPromark]);

  // move doLayersContainsVector and setVector in effect to avoid heavy render process
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(() => {
    const newVal = doLayersContainsVector(selectedLayers);
    if (hasVector !== newVal) {
      setHasVector(newVal);
    }
    if (isPromark) {
      const newGradientVal = selectedLayers.some((layerName: string) => {
        const layer = getLayerElementByName(layerName);
        return !!layer?.querySelector('image[data-shading="true"]');
      });
      if (newGradientVal !== hasGradient) setHasGradient(newGradientVal);
    }
  });

  return (
    <LayerPanelContext.Provider
      value={{
        selectedLayers,
        setSelectedLayers: lazySetSelectedLayers,
        forceUpdate,
        forceUpdateSelectedLayers,
        hasVector,
        hasGradient,
      }}
    >
      {children}
    </LayerPanelContext.Provider>
  );
};
