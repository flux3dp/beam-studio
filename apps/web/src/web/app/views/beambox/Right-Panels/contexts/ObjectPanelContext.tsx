import React, { useCallback, useEffect, useRef, useState } from 'react';

import eventEmitterFactory from 'helpers/eventEmitterFactory';
import useForceUpdate from 'helpers/use-force-update';

interface IObjectPanelContext {
  polygonSides: number;
  dimensionValues: any;
  activeKey: string | null;
  updateDimensionValues: (newValues: any) => void;
  getDimensionValues: (
    response: {
      dimensionValues: any;
    },
    key?: string
  ) => any;
  updateActiveKey: (activeKey: string | null) => void;
  updateObjectPanel: () => void;
}

export const ObjectPanelContext = React.createContext<IObjectPanelContext>({
  polygonSides: 5,
  dimensionValues: {},
  activeKey: null,
  updateDimensionValues: () => {},
  getDimensionValues: () => {},
  updateActiveKey: () => {},
  updateObjectPanel: () => {},
});
const objectPanelEventEmitter = eventEmitterFactory.createEventEmitter('object-panel');
const minRenderInterval = 200;

interface Props {
  children?: React.ReactNode;
}

export const ObjectPanelContextProvider = ({ children }: Props): JSX.Element => {
  const forceUpdate = useForceUpdate();
  const [polygonSides, setPolygonSides] = useState(5);
  const [activeKey, setActiveKey] = useState<string | null>(null);
  const dimensionValues = useRef<{ [key: string]: number | string }>({});
  const lastUpdateTime = useRef(Date.now());
  const updateTimeout = useRef<NodeJS.Timeout>(null);
  useEffect(() => {
    objectPanelEventEmitter.on('UPDATE_POLYGON_SIDES', setPolygonSides);
    objectPanelEventEmitter.on('UPDATE_ACTIVE_KEY', setActiveKey);
    return () => {
      objectPanelEventEmitter.removeListener('UPDATE_POLYGON_SIDES', setPolygonSides);
      objectPanelEventEmitter.removeListener('UPDATE_ACTIVE_KEY', setActiveKey);
    }
  }, []);

  const getDimensionValues = useCallback((response: { dimensionValues: any }, key?: string) => {
    response.dimensionValues = key ? dimensionValues.current[key] : dimensionValues.current;
  }, []);

  useEffect(() => {
    objectPanelEventEmitter.on('GET_DIMENSION_VALUES', getDimensionValues);
    return () => {
      objectPanelEventEmitter.removeListener('GET_DIMENSION_VALUES', getDimensionValues);
    }
  }, [getDimensionValues]);
  const updateDimensionValues = useCallback((newValues: { [key: string]: number | string }) => {
    dimensionValues.current = {
      ...dimensionValues.current,
      ...newValues,
    };
  }, []);

  useEffect(() => {
    objectPanelEventEmitter.on('UPDATE_DIMENSION_VALUES', updateDimensionValues);
    return () => {
      objectPanelEventEmitter.removeListener('UPDATE_DIMENSION_VALUES', updateDimensionValues);
    }
  }, [updateDimensionValues]);
  const getActiveKey = useCallback((response: { activeKey: string | null }) => {
    response.activeKey = activeKey;
  }, [activeKey]);

  useEffect(() => {
    objectPanelEventEmitter.on('GET_ACTIVE_KEY', getActiveKey);
    return () => {
      objectPanelEventEmitter.removeListener('GET_ACTIVE_KEY', getActiveKey);
    }
  }, [getActiveKey]);

  const updateObjectPanel = useCallback(() => {
    clearTimeout(updateTimeout.current);
    const time = Date.now();
    if (time - lastUpdateTime.current >= minRenderInterval) {
      lastUpdateTime.current = time;
      forceUpdate();
    } else {
      updateTimeout.current = setTimeout(() => {
        lastUpdateTime.current += minRenderInterval;
        forceUpdate();
      }, lastUpdateTime.current + minRenderInterval - time);
    }
  }, [forceUpdate]);
  useEffect(() => {
    objectPanelEventEmitter.on('UPDATE_OBJECT_PANEL', updateObjectPanel);
    return () => {
      objectPanelEventEmitter.removeListener('UPDATE_OBJECT_PANEL', updateObjectPanel);
    }
  }, [updateObjectPanel]);

  return (
    <ObjectPanelContext.Provider
      value={{
        polygonSides,
        dimensionValues: dimensionValues.current,
        activeKey,
        updateDimensionValues,
        getDimensionValues,
        updateActiveKey: setActiveKey,
        updateObjectPanel,
      }}
    >
      {children}
    </ObjectPanelContext.Provider>
  );
};
