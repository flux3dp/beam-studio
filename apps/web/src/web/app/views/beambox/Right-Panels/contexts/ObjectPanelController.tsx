import eventEmitterFactory from 'helpers/eventEmitterFactory';

const objectPanelEventEmitter = eventEmitterFactory.createEventEmitter('object-panel');

const updateDimensionValues = (newValue: any): void => {
  objectPanelEventEmitter.emit('UPDATE_DIMENSION_VALUES', newValue);
};

const getDimensionValues = (key?: string): any => {
  const response = {
    dimensionValues: {},
  };
  objectPanelEventEmitter.emit('GET_DIMENSION_VALUES', response, key);
  return response.dimensionValues;
};

const minEventInterval = 50;
let updateObjectPanelTimeout = null;
const updateObjectPanel = (): void => {
  clearTimeout(updateObjectPanelTimeout);
  updateObjectPanelTimeout = setTimeout(() => {
    objectPanelEventEmitter.emit('UPDATE_OBJECT_PANEL');
  }, minEventInterval);
};

const updatePolygonSides = (polygonSides: number): void => {
  objectPanelEventEmitter.emit('UPDATE_POLYGON_SIDES', polygonSides);
};

const updateActiveKey = (activeKey: string | null): void => {
  objectPanelEventEmitter.emit('UPDATE_ACTIVE_KEY', activeKey);
};

const getActiveKey = (): string | null => {
  const response = {
    activeKey: undefined,
  };
  objectPanelEventEmitter.emit('GET_ACTIVE_KEY', response);
  return response.activeKey;
};

export default {
  updateObjectPanel,
  updateDimensionValues,
  getDimensionValues,
  updatePolygonSides,
  updateActiveKey,
  getActiveKey,
  events: objectPanelEventEmitter,
};
