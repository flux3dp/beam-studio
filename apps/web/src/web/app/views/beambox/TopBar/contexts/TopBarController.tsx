import BeamboxGlobalInteraction from 'app/actions/beambox/beambox-global-interaction';
import eventEmitterFactory from 'helpers/eventEmitterFactory';
import { IDeviceInfo } from 'interfaces/IDevice';

const topBarEventEmitter = eventEmitterFactory.createEventEmitter('top-bar');

const setElement = (elem: Element | null): void => {
  if (!elem) {
    BeamboxGlobalInteraction.onObjectBlur();
  } else {
    BeamboxGlobalInteraction.onObjectBlur();
    BeamboxGlobalInteraction.onObjectFocus([elem]);
  }

  topBarEventEmitter.emit('SET_ELEMENT', elem);
};

const updateTitle = (title: string, isCloudFile: boolean): void => {
  topBarEventEmitter.emit('UPDATE_TITLE', title, isCloudFile);
};

const onTitleChange = (handler: (title: string, isCloudFile: boolean) => void): void => {
  topBarEventEmitter.on('UPDATE_TITLE', handler);
};

const offTitleChange = (handler: (title: string, isCloudFile: boolean) => void): void => {
  topBarEventEmitter.removeListener('UPDATE_TITLE', handler);
};

const setHasUnsavedChange = (hasUnsavedChange: boolean): void => {
  topBarEventEmitter.emit('SET_HAS_UNSAVED_CHANGE', hasUnsavedChange);
};

const getTopBarPreviewMode = (): boolean => {
  const response = {
    isPreviewMode: false,
  };
  topBarEventEmitter.emit('GET_TOP_BAR_PREVIEW_MODE', response);
  return response.isPreviewMode;
};

const getSelectedDevice = (): IDeviceInfo | null => {
  const response = {
    selectedDevice: null,
  };
  topBarEventEmitter.emit('GET_SELECTED_DEVICE', response);
  return response.selectedDevice;
};

const setSelectedDevice = (device: IDeviceInfo): void => {
  topBarEventEmitter.emit('SET_SELECTED_DEVICE', device);
};

export default {
  setElement,
  updateTitle,
  setHasUnsavedChange,
  getTopBarPreviewMode,
  getSelectedDevice,
  setSelectedDevice,
  onTitleChange,
  offTitleChange,
};
