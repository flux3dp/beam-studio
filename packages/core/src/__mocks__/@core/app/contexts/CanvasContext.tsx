import { createContext } from 'react';

export const CanvasContext = createContext({});
