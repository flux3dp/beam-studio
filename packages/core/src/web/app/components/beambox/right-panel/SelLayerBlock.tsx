import classNames from 'classnames';
import React, { memo, useContext, useEffect, useState } from 'react';

import useI18n from 'helpers/useI18n';
import { getObjectLayer, moveToOtherLayer } from 'helpers/layer/layer-helper';
import { LayerPanelContext } from 'app/views/beambox/Right-Panels/contexts/LayerPanelContext';
import { SelectedElementContext } from 'app/contexts/SelectedElementContext';

import styles from './SelLayerBlock.module.scss';

const defaultOption = ' ';

interface Props {
  layerNames: string[];
}

function SelLayerBlock({ layerNames }: Props): JSX.Element {
  const lang = useI18n().beambox.right_panel.layer_panel;
  const [promptMoveLayerOnce, setPromptMoveLayerOnce] = useState(false);
  const [displayValue, setDisplayValue] = useState(defaultOption);
  const { selectedElement } = useContext(SelectedElementContext);
  const { selectedLayers } = useContext(LayerPanelContext);
  useEffect(() => {
    if (!selectedElement) return;
    if (selectedElement.getAttribute('data-tempgroup') === 'true') {
      const originalLayers = new Set(
        ([...selectedElement.childNodes] as SVGElement[])
          .filter((elem) => elem?.getAttribute('data-imageborder') !== 'true')
          .map((elem) => elem.getAttribute('data-original-layer'))
      );
      if (originalLayers.size === 1) {
        const [firstValue] = originalLayers;
        setDisplayValue(firstValue ?? defaultOption);
      } else setDisplayValue(defaultOption);
    } else {
      const currentLayer = getObjectLayer(selectedElement as SVGElement);
      const currentLayerName = currentLayer?.title ?? defaultOption;
      setDisplayValue(currentLayerName);
    }
  }, [selectedElement, selectedLayers]);
  if (!selectedElement) return null;
  if (layerNames.length === 1) return null;

  const onChange = (e: React.ChangeEvent) => {
    const select = e.target as HTMLSelectElement;
    const destLayer = select.options[select.selectedIndex].value;
    moveToOtherLayer(
      destLayer,
      () => {
        setDisplayValue(destLayer);
        setPromptMoveLayerOnce(true);
      },
      !promptMoveLayerOnce
    );
  };

  const options = [];
  if (displayValue === defaultOption) {
    options.push(
      <option value={defaultOption} key={-1}>
        {lang.move_elems_to}
      </option>
    );
  }
  for (let i = layerNames.length - 1; i >= 0; i -= 1) {
    const layerName = layerNames[i];
    options.push(
      <option value={layerName} key={i}>
        {layerName}
      </option>
    );
  }
  return (
    <div className={classNames('controls', styles.container)}>
      <span className={styles.label}>{lang.move_elems_to}</span>
      <select
        className={styles.select}
        value={displayValue}
        title="Move selected elements to a different layer"
        onChange={onChange}
        disabled={options.length < 2}
      >
        {options}
      </select>
    </div>
  );
}

export default memo(SelLayerBlock);
