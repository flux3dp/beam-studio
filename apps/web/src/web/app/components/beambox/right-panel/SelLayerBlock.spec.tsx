import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import { SelectedElementContext } from 'app/contexts/SelectedElementContext';

import SelLayerBlock from './SelLayerBlock';

jest.mock('helpers/useI18n', () => () => ({
  beambox: {
    right_panel: {
      layer_panel: {
        move_elems_to: 'move_elems_to',
      },
    },
  },
}));

const mockgetObjectLayer = jest.fn();
const mockMoveToOtherLayer = jest.fn();
jest.mock('helpers/layer/layer-helper', () => ({
  moveToOtherLayer: (...args: any[]) => mockMoveToOtherLayer(...args),
  getObjectLayer: (...args: any[]) => mockgetObjectLayer(...args),
}));

describe('SelLayerBlock', () => {
  it('should render correctly when layer is less than 1', () => {
    const { container } = render(<SelLayerBlock layerNames={['layer1']} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when layer is more than 1 and selected elemented is none', () => {
    const { container } = render(<SelLayerBlock layerNames={['layer1', 'layer2']} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when layer is more than 1 and selected elemented is not in a temp group', () => {
    mockgetObjectLayer.mockReturnValue({ title: 'layer1' });
    const mockElem = { getAttribute: () => jest.fn() };
    const { container } = render(
      <SelectedElementContext.Provider value={{ selectedElement: mockElem } as any}>
        <SelLayerBlock layerNames={['layer1', 'layer2']} />
      </SelectedElementContext.Provider>
    );
    expect(mockgetObjectLayer).toBeCalledTimes(1);
    expect(mockgetObjectLayer).toHaveBeenLastCalledWith(mockElem);
    expect(container).toMatchSnapshot();
    const select = container.querySelector('select') as HTMLSelectElement;
    expect(select.value).toBe('layer1');
  });

  test('move to other layer', () => {
    mockgetObjectLayer.mockReturnValue({ title: 'layer1' });
    const mockElem = { getAttribute: () => jest.fn() };
    const { container } = render(
      <SelectedElementContext.Provider value={{ selectedElement: mockElem } as any}>
        <SelLayerBlock layerNames={['layer1', 'layer2']} />
      </SelectedElementContext.Provider>
    );
    const select = container.querySelector('select') as HTMLSelectElement;
    fireEvent.change(select, { target: { value: 'layer2' } });
    expect(mockMoveToOtherLayer).toBeCalledTimes(1);
    expect(mockMoveToOtherLayer).toHaveBeenLastCalledWith(
      'layer2',
      expect.any(Function),
      true
    );
  });
});
