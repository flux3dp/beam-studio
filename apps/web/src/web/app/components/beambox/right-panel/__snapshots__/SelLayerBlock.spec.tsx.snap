// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelLayerBlock should render correctly when layer is less than 1 1`] = `<div />`;

exports[`SelLayerBlock should render correctly when layer is more than 1 and selected elemented is none 1`] = `<div />`;

exports[`SelLayerBlock should render correctly when layer is more than 1 and selected elemented is not in a temp group 1`] = `
<div>
  <div
    class="controls container"
  >
    <span
      class="label"
    >
      move_elems_to
    </span>
    <select
      class="select"
      title="Move selected elements to a different layer"
    >
      <option
        value="layer2"
      >
        layer2
      </option>
      <option
        value="layer1"
      >
        layer1
      </option>
    </select>
  </div>
</div>
`;
